src/Microsoft.AspNetCore.Http.Features/FeatureReferences.cs

```diff
@@ -26,34 +26,53 @@ public FeatureReferences(IFeatureCollection collection)
         public TFeature Fetch<TFeature, TState>(
             ref TFeature cached,
             TState state,
-            Func<TState, TFeature> factory)
+            Func<TState, TFeature> factory) where TFeature : class
         {
-            var cleared = false;
-            if (Revision != Collection.Revision)
+            if (Revision == 0)
```

Comment:
If `cached == null` it couldn't update the revision count as it may have changed and other cached features may be invalid; which would mean on second use it would have to clear the cache.